src/lsm/README.md

```diff
@@ -0,0 +1,111 @@
+# Tree
+## Compaction
+
+Tree compaction runs to the sound of music!
+
+Compacting LSM trees involves merging and moving tables into the next levels as needed.
+To avoid write amplification stalls and bound latency, compaction is done incrementally.
+
+A full compaction phase is denoted as a bar or measure, using terms from music notation.
+Each bar consists of `lsm_batch_multiple` beats or "compaction ticks" of work.
+A compaction beat is started asynchronously with `compact()` which takes a callback and
+the op that was just committed.
+
+A bar is split in half according to the "first" down beat and "middle" down beat.
+The first half of the bar compacts even levels while the latter compacts odd levels.
+Mutable table changes are sorted and compacted into the immutable table.
+The immutable table is compacted into level 0 during the odd level half of the bar.
+
+At any given point, there are at most levels/2 compactions running concurrently.
+The source level is denoted as `level_a` and the target level as `level_b`.
+The last level in the LSM tree has no target level so it is never a source level.
+
+Assuming a bar/`lsm_batch_multiple` of 4, the invariants can be described as follows:
+
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
```

Comment:
How do tables and levels relate? Does table contain a list of pointers to levels on disk?

Reply:
I will need to write something up on what a manifest is, but that is probably out of scope for this PR. Each table lives in a single level. The manifest is an index that tracks the location (i.e. the level) of every table, along with other table metadata.

Reply:
Also out of scope, but could you put a vocab section at the top of the readme? Could maybe include some of lsm, forest, groove, tree, table (mutable/immutable), level, manifest, beat, bar, measure (same as a bar?), snapshot.

Or I could write it in a separate PR, but I'll have to ping you for some of them.